src/Microsoft.ML.OnnxTransform/OnnxTransform.cs

```diff
@@ -129,33 +154,52 @@ private OnnxTransform(IHostEnvironment env, Arguments args, byte[] modelBytes =
                 Model = OnnxModel.CreateFromBytes(modelBytes);
 
             var modelInfo = Model.ModelInfo;
-            if (modelInfo.InputsInfo.Length != 1)
-                throw env.Except($"OnnxTransform supports Onnx models with one input. The provided model has ${modelInfo.InputsInfo.Length} input(s).");
-            if (modelInfo.OutputsInfo.Length != 1)
-                throw env.Except($"OnnxTransform supports Onnx models with one output. The provided model has ${modelInfo.OutputsInfo.Length} output(s).");
-
-            Input = args.InputColumn;
-            Output = args.OutputColumn;
-
-            var outputNodeInfo = Model.ModelInfo.OutputsInfo[0];
-            var type = OnnxUtils.OnnxToMlNetType(outputNodeInfo.Type);
-            var shape = outputNodeInfo.Shape;
-            var dims = shape.Count > 0 ? shape.Skip(shape[0] < 0 ? 1 : 0).Select( x => (int) x ).ToArray() : new[] { 0 };
-            OutputType = new VectorType(type, dims);
+            //if (modelInfo.InputsInfo.Length != 1)
+            //    throw env.Except($"OnnxTransform supports Onnx models with one input. The provided model has ${modelInfo.InputsInfo.Length} input(s).");
+            //if (modelInfo.OutputsInfo.Length != 1)
+            //    throw env.Except($"OnnxTransform supports Onnx models with one output. The provided model has ${modelInfo.OutputsInfo.Length} output(s).");
+
+            Inputs = args.InputColumns;
+            Outputs = args.OutputColumns;
+            //var type = OnnxUtils.OnnxToMlNetType(outputNodeInfo.Type);
+            //var shape = outputNodeInfo.Shape;
+            //var dims = shape.Count > 0 ? shape.Skip(shape[0] < 0 ? 1 : 0).Select( x => (int) x ).ToArray() : new[] { 0 };
+
+            OutputTypes = new ColumnType[args.OutputColumns.Length];
+
+            var numModelOutputs = Model.ModelInfo.OutputsInfo.Length;
+            for (int i=0; i < args.OutputColumns.Length; i++)
+            {
+                var idx = -1;
+                for (var j = 0; j < Model.ModelInfo.OutputsInfo.Length; j++)
+                    if (Model.ModelInfo.OutputsInfo[j].Name == args.OutputColumns[i])
+                    {
+                        idx = j;
+                        break;
+                    }
+                if (idx < 0)
+                    throw _host.Except($"Column {args.OutputColumns[i]} doesn't match output node names of model");
+                var outputNodeInfo = Model.ModelInfo.OutputsInfo[idx];
+                var shape = outputNodeInfo.Shape;
+                var dims = shape.Count > 0 ? shape.Skip(shape[0] < 0 ? 1 : 0).Select(x => (int)x).ToArray() : new[] { 0 };
```

Comment:
>var dims = shape.Count > 0 ? shape.Skip(shape[0] < 0 ? 1 : 0).Select(x => (int)x).ToArray() : new[] { 0 }; [](start = 16, length = 106)

This is too dense for me to follow. Please unpack the code into different cases and add comments about each case (meaning of negative count, why we skip, why skip 0 or 1, etc).

I've seen this later in the code to. If makes sense, please refactor to separate method.

Reply:
refactored

---
In reply to: [232836841](https://github.com/dotnet/machinelearning/pull/1586#discussion_r232836841) [](ancestors = 232836841)